design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+    The provider that is built from the source code in this repository can be installed into a Crossplane control plane and adds the following new functionality:
+
+    * Custom Resource Definitions (CRDs) that model GCP infrastructure and services (e.g. [CloudSQL](https://cloud.google.com/sql/), [GKE clusters](https://cloud.google.com/kubernetes-engine/), etc.)
+    * Controllers to provision these resources in GCP based on the users desired state captured in CRDs they create
+    * Implementations of Crossplane's portable resource abstractions, enabling GCP resources to fulfill a user's general need for cloud services
+  category: Infrastructure
+  version: 0.1.0
+  icons:
+  - base64data: iVBORw0KGgoAAAANSUhEUgAAAOEAAADZCAYAAADWmle6AAAACXBIWXMA
+    mediatype: image/png
+  maintainers:
+  - name: Nic Cope
+   email: nic@upbound.io
+  owners:
+  - name: Nic Cope
+    email: nic@upbound.io
+  keywords:
+  - "google"
+  - "gcp"
+  - "infrastructure"
+  website: "https://crossplane.io"
+  controller:
+    deployment:
+      name: provider-gcp-controller
+      spec:
+        replicas: 1
+        selector:
+          matchLabels:
+            core.crossplane.io/name: "provider-gcp"
+        template:
+          metadata:
+            name: provider-gcp-controller
+            labels:
+              core.crossplane.io/name: "provider-gcp"
+          spec:
+            serviceAccountName: provider-gcp-controller
+            containers:
+            - name: provider-gcp-controller
+              image: crossplane/provider-gcp:v0.10.0
+              imagePullPolicy: Always
+```
+
+User creates `Package`:
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: Package
+metadata:
+  name: secure-stacl
+spec:
+  package: crossplane/secure-stack:v0.1.0
+```
+
+This `Package` only installs composition types (no controller) and depends on
+`provider-gcp`:
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: PackageRevision
+metadata:
+  name: cc8a4eb3a83721817ac76c7080622aec8c999538230db98be9755fea66e6982d
+  ownerReferences:
+  - apiVersion: packages.crossplane.io/v1alpha1
+    controller: true
+    blockOwnerDeletion: true
+    kind: Package
+    name: secure-stack
+    uid: d9607e19-f88f-11e6-a518-42010a800195
+spec:
+  infrastructureDefinitions:
+  - name: network.securestack.crossplane.io
+  - name: postgres.securestack.crossplane.io
+  - name: mysql.securestack.crossplane.io
+  - name: cluster.securestack.crossplane.io
+  - name: vm.securestack.crossplane.io
+  ...
+  compositions:
+  - name: network.gcp.securestack.crossplane.io
+  - name: network.aws.securestack.crossplane.io
+  - name: network.azure.securestack.crossplane.io
+  - name: network.alibaba.securestack.crossplane.io
+  - name: postgres.gcp.securestack.crossplane.io
+  - name: postgres.aws.securestack.crossplane.io
+  - name: postgres.azure.securestack.crossplane.io
+  - name: postgres.alibaba.securestack.crossplane.io
+  ...
+  dependsOn:
```

Comment:
I am trying to think about how our dependency model will work, and it seems like there are two main cases:

- Installation of a `Provider` that manages CRDs that it does not install itself
- Installation of a `Composition` that composes managed resources not installed by the same package

The issue with basing dependencies on CRD GVK is that just because a CRD exists with the correct GVK, we cannot guarantee that it is actually the one that the package being installed depends on, and that it is being managed by the controller it expects. For this reason, it seems to make more sense to say that packages can only declare dependencies on other packages, and they specify by hash. For instance, if testing a "stack" that introduced composition types that depend on `provider-gcp`, we do not just need to know that `vpc.network.gcp.crossplane.io/v1beta1` exists, but that it is the one we tested against, which can only be verified if we declare the package hash of the version we used. This is at odds with our desire to do "dynamic CRD resolution".

Reply:
cc @negz @jbw976

Reply:
> it seems to make more sense to say that packages can only declare dependencies on other packages, and they specify by hash

IMO, that approach would have the downside of a declared dependency needing to be updated for every new hash of the package it depends, regardless of if anything functional has changed.  e.g.,

* package A depends on package B:hashB
* package B gets a tiny bug fix to its controller logic in a new version B:hashB2
* package A would have to update it's declared dependency to say it now depends on B:hashB2 if it wasn't to get the latest package B installed via its dependency

That doesn't sound sustainable or very smooth, it puts a requirement on packages to update their declared dependencies all the time.

I'm hoping that when GVK's are used appropriately, this dependency resolution works out better.  In the same example:

* package A depends on a CRD in package B with GVK v1beta1
* package B gets a tiny bug fix to its controller logic in a new version B:hashB2
* package A can still be installed, with no updates to its dependencies, and get the latest package B controller fix.

When there are changes to the CRD schema that would affect other packages that depend on it, a package MUST update it's GVK for the affected types.  I wrote a bit about this (but didn't solve everything haha) in the first iteration of an older design doc: https://github.com/crossplane/crossplane/commit/e9df5dee5f4972cd931fde5a9ad3ad9b88e177a5#diff-8989bf194102b27431a10a02d42afd96R194-R196

My gut is that I would choose the flexibility of depending on GVKs instead of package hashes, with the assumption that package authors have to be diligent and thoughtful with their GVK versioning practices.

Reply:
@jbw976 that is certainly a valid concern, and we would need to find out how to address doing updates in respect to dependencies without making it a cumbersome process for package authors. However, this isn't really a question of proper versioning of CRDs within a package (i.e. Package B making small changes and properly updating any versions of its CRDs if required). This is a question of whether `rdsinstance.aws.crossplane.io/v1alpha1` in Package B is the CRD that Package A is expecting or `rdsinstance.aws.crossplane.io/v1alpha1` in Package C is the CRD that Package A is expecting. Because packages are not distinguished by the CRDs they bring (but rather by an OCI image), there is no way to guarantee which CRD is the "true" one. There is a way to guarantee the "true" package though using image hash.

If an image registry was purpose-built to validate a Crossplane package, then it could restrict packages ever introducing the same CRD as another package. For instance, on docker hub I can push my image as `hasheddan/cool-image:v0.0.1` and no one else can do the same. When someone pulls from `hub.docker.com/hasheddan/cool-image:v0.0.1` they can know what they are getting (especially if they also provide the hash). However, in Crossplane we are trying to verify uniqueness of the contents of that image (i.e. the CRDs), which is not possible by just specifying the GVK.

One option could be to say that a registry for Crossplane images should only let you push images (packages) with CRDs in your username or org (i.e. I could only push packages with CRDs in the hasheddan.io group), and if the registry does not perform that validation then all bets are off.

Reply:
Yes I think that's a valid concern to have and I definitely think that a registry that supports CRD -> Package lookups (by indexing CRDs in each package) should also probably support enforcement of only 1 package being the definer or owner of a given CRD.  Multiple versions of package A could define a CRD, but package B should not also be allowed to define the same CRD.

My gut feel is still to avoid forcing package authors from specifying their dependency to be a specific hash of a package, that just feels too rigid and inflexible. Hopefully by making an assumption about the existence of a registry that can do CRD -> Package lookups, this issue is less of a concern.

Reply:
> The issue with basing dependencies on CRD GVK is that just because a CRD exists with the correct GVK, we cannot guarantee that it is actually the one that the package being installed depends on, and that it is being managed by the controller it expects. For this reason, it seems to make more sense to say that packages can only declare dependencies on other packages, and they specify by hash.

I couldn't understand the problem that the hash solves here. As a package author, I can write a dependency as:
```yaml
dependsOn:
- crossplane/provider-aws:0.10/vpc.ec2.aws.crossplane.io/v1beta1
```
or 
```yaml
- source: crossplane/provider-aws:0.10
  crds:
  - vpc.ec2.aws.crossplane.io/v1beta1
```

At this point, I have identified an exact controller in an image. If, for some reason, `crossplane/provider-aws:0.10` resolves to another image that I had in mind then that's maybe what they want, i.e. `replace` statement of `go.mod` where you don't touch the package itself but change how it's resolved to get your own fork for example. Package author having the ability to pin me down to an exact hash feels kind of restricting.